lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1127,6 +1127,78 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         }
       })
 
+      // Handle memories.
+      .Case<MemOp>([&](auto op) {
+        // Create constraint variables for all ports.
+        for (auto result : op.results())
+          declareVars(result, op.getLoc());
+
+        // This creates independent variables for every data port. Yet, what we
+        // actually want is for all data ports to share the same variable. To do
+        // this, we find the first data port declared, and use that port's vars
+        // for all the other ports.
+        LLVM_DEBUG(llvm::dbgs() << "Adjusting memory port variables:\n");
+        auto first = op.getResult(0);
+        auto firstType =
+            op.getPortType(0).getPassiveType().template cast<BundleType>();
+        auto firstKind = op.getPortKind(0);
+        unsigned firstBaseID;
+        switch (firstKind) {
+        case MemOp::PortKind::Read:
+          firstBaseID = firstType.getFieldID(3);
+          break;
+        case MemOp::PortKind::Write:
+          firstBaseID = firstType.getFieldID(3);
+          break;
+        case MemOp::PortKind::ReadWrite:
+          firstBaseID = firstType.getFieldID(4);
+          break;
+        }
+        unsigned firstMaxID = op.getDataType().getMaxFieldID();
+        LLVM_DEBUG(llvm::dbgs()
+                   << "- Data type max field ID: " << firstMaxID << "\n");
+        LLVM_DEBUG(llvm::dbgs() << "- First port type: " << firstType << "\n");
+        LLVM_DEBUG(llvm::dbgs()
+                   << "- First port base ID: " << firstBaseID << "\n");
+
+        // Reuse data port variables.
+        for (unsigned i = 0, e = op.results().size(); i < e; ++i) {
+          auto result = op.getResult(i);
+          auto reuseVars = [&](unsigned baseID) {
+            for (unsigned i = 0; i <= firstMaxID; ++i) {
+              auto srcField = FieldRef(first, firstBaseID + i);
+              auto dstField = FieldRef(result, baseID + i);
+              if (srcField == dstField)
+                continue;
+              LLVM_DEBUG(llvm::dbgs()
+                         << "- Reusing " << getFieldName(srcField) << " ("
+                         << srcField.getFieldID() << ") for "
+                         << getFieldName(dstField) << " ("
+                         << dstField.getFieldID() << ")\n");
+              if (auto *e = getExprOrNull(srcField)) {
+                setExpr(dstField, e);
+                assert(getExpr(dstField) == e);
+              }
+            }
+          };
```

Comment:
1. Should it start on `i=1`, and then you can remove the `if (srcField == dstField)` check?
2. It might be possible to do something with `unifyTypes`, extend it to take two `FieldRefs` and a `Type` instead of calling this lamba.  I'm not sure if it would be better: the upside is you won't try every possibly field id and look up a non-inserted expr for a field ref, the downside is it uses a a rec lambda+typeswitch.  Another option is to just make `unifyTypes` use the same algo here if its faster.

Reply:
On (1): I think I can move the check out of the loop and use the firstBaseID and baseID for the check. I'm trying to skip the very first port (tricky because readwrite have two data ports).

Reply:
On (2): Sounds great! I'll take a look!

Reply:
Awesome suggestions @youngar. I updated the `unifyTypes` function to take two `FieldRef`s, which allowed me to use it here as well. :+1: